app/src/full/java/io/homeassistant/companion/android/sensors/HealthConnectSensorManager.kt

```diff
@@ -216,13 +312,179 @@ class HealthConnectSensorManager : SensorManager {
+        val elevationGainedRequest = ReadRecordsRequest(
+            recordType = ElevationGainedRecord::class,
+            timeRangeFilter = TimeRangeFilter.between(
+                Instant.now().minus(30, ChronoUnit.DAYS),
+                Instant.now()
+            ),
+            ascendingOrder = false,
+            pageSize = 1
+        )
+        val response = runBlocking { healthConnectClient.readRecords(elevationGainedRequest) }
+        if (response.records.isEmpty()) {
+            return
+        }
+        onSensorUpdated(
+            context,
+            elevationGained,
+            response.records.last().elevation.inMeters,
+            elevationGained.statelessIcon,
+            attributes = mapOf(
+                "endTime" to response.records.last().endTime,
+                "source" to response.records.last().metadata.dataOrigin.packageName
+            )
+        )
+    }
+
+    private fun updateFloorsClimbedSensor(context: Context) {
+        val healthConnectClient = getOrCreateHealthConnectClient(context) ?: return
+        val floorsClimbedRequest = ReadRecordsRequest(
+            recordType = FloorsClimbedRecord::class,
+            timeRangeFilter = TimeRangeFilter.between(
+                Instant.now().minus(30, ChronoUnit.DAYS),
+                Instant.now()
+            ),
+            ascendingOrder = false,
+            pageSize = 1
+        )
+        val response = runBlocking { healthConnectClient.readRecords(floorsClimbedRequest) }
+        if (response.records.isEmpty()) {
+            return
+        }
+        onSensorUpdated(
+            context,
+            floorsClimbed,
+            response.records.last().floors,
+            floorsClimbed.statelessIcon,
+            attributes = mapOf(
+                "endTime" to response.records.last().endTime,
+                "source" to response.records.last().metadata.dataOrigin.packageName
+            )
+        )
+    }
+
+    private fun updateSleepDurationSensor(context: Context) {
+        val healthConnectClient = getOrCreateHealthConnectClient(context) ?: return
+        val sleepRequest = runBlocking {
+            healthConnectClient.aggregate(
+                AggregateRequest(
+                    metrics = setOf(SleepSessionRecord.SLEEP_DURATION_TOTAL),
+                    timeRangeFilter = TimeRangeFilter.between(
+                        LocalDateTime.of(LocalDate.now(), LocalTime.MIDNIGHT),
+                        LocalDateTime.of(LocalDate.now(), LocalTime.now())
+                    )
+                )
+            )
+        }
+        sleepRequest[SleepSessionRecord.SLEEP_DURATION_TOTAL]?.let { sleep ->
+            onSensorUpdated(
+                context,
+                sleepDuration,
+                sleep.toMinutes(),
+                sleepDuration.statelessIcon,
+                attributes = mapOf(
+                    "endTime" to Instant.now(),
```

Comment:
Just to be sure: have you considered to use the end of the sleep session instead of now? If it is for example 15:00 and I've slept 23:00-7:30 seeing the end time of my sleep duration as 15:00 doesn't make a lot of sense.

Reply:
this now pulls from the last record so should be more precise :)